File: qrs-core-python/src/qrs/services/certificateService.py
"""CertificateService: key generation and TCert creation.

A TCert combines an issuer identity and the schema of one document type. It is
always self-signed, and its ``key_id`` is derived from its public key. A single
key pair may own several TCerts (one per document type), distinguished by their
certificate number.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..cbor import cbor_encode
from ..crypto.providers import KeyPairMaterial
from ..deps import ServiceDeps
from ..envelope import (
    build_signed_object,
    parse_signed_object,
    tcert_id_of,
    verify_parsed_signed_object,
)
from ..errors import QrsCryptoError, QrsNotFoundError, QrsValidationError
from ..fields import FieldSchema
from ..id import from_hex, is_hash_algorithm, to_hex
from ..signed_object import assert_valid_object_data, is_field_type

__all__ = ["CertificateService", "CreateTcertParams", "CreateTcertResult"]


@dataclass
class CreateTcertParams:
    algorithm: str
    name: str
    fields: list[FieldSchema]
    key_id: str | None = None
    valid_after: int | None = None
    valid_before: int | None = None
    sdoc_max_age_seconds: int | None = None
    hash_algorithm: str | None = None
    online_endpoint: str | None = None
    metadata: dict[str, Any] | None = None


@dataclass
class CreateTcertResult:
    key_id: str
    tcert_id: str
    certificate_number: int
    bytes: bytes
    parsed: Any


class CertificateService:
    def __init__(self, deps: ServiceDeps) -> None:
        self._deps = deps

    async def generate_key_pair(self, algorithm: str) -> str:
        provider = self._deps.crypto_registry.get(algorithm)
        pair = provider.generate_key_pair()
        key_id = provider.key_id(pair.public_jwk)
        await self._deps.private_key_store.save(key_id, algorithm, pair.private_jwk)
        await self._deps.public_key_store.save(key_id, algorithm, pair.public_jwk)
        return key_id

    async def create_tcert(self, params: CreateTcertParams) -> CreateTcertResult:
        for field in params.fields:
            if not is_field_type(field.type):
                raise QrsValidationError(f"Unsupported field type: {field.type}")
            if not field.name or not field.label:
                raise QrsValidationError("Each field must have a name and a label")
            if not self._deps.field_registry.has(field.type):
                raise QrsValidationError(f"No field engine registered for type: {field.type}")

        key_id = params.key_id
        if not key_id:
            key_id = await self.generate_key_pair(params.algorithm)

        pub_rec = await self._deps.public_key_store.load(key_id)
        if not pub_rec:
            raise QrsNotFoundError(f"Public key not found: {key_id}")
        priv_rec = await self._deps.private_key_store.load(key_id)
        if not priv_rec:
            raise QrsNotFoundError(f"Private key not available for signing: {key_id}")

        provider = self._deps.crypto_registry.get(pub_rec["algorithm"])

        # Pick the lowest unused certificate number (1..255) for this key.
        existing = await self._deps.certificate_store.find_by_key_id(key_id)
        used = {int(tcert_id.split(":")[1]) for tcert_id, _ in existing}
        certificate_number = 0
        for n in range(1, 256):
            if n not in used:
                certificate_number = n
                break
        if certificate_number == 0:
            raise QrsValidationError("No certificate numbers left for this key")

        data: dict[str, Any] = {
            "keyId": from_hex(key_id),
            "certificateNumber": certificate_number,
            "algorithm": provider.algorithm,
            "publicKey": pub_rec["public_jwk"],
            "identity": {"name": params.name},
            "schema": [f.to_json() for f in params.fields],
        }
        validity: dict[str, Any] = {}
        if params.valid_after is not None:
            validity["validAfter"] = params.valid_after
        if params.valid_before is not None:
            validity["validBefore"] = params.valid_before
        if params.sdoc_max_age_seconds is not None:
            validity["sdocMaxAgeSeconds"] = params.sdoc_max_age_seconds
        if validity:
            data["validity"] = validity
        if params.hash_algorithm is not None:
            if not is_hash_algorithm(params.hash_algorithm):
                raise QrsValidationError(f"Unsupported hash algorithm: {params.hash_algorithm}")
            data["hashAlgorithm"] = params.hash_algorithm
        if params.online_endpoint:
            data["onlineEndpoint"] = params.online_endpoint
        if params.metadata:
            data["metadata"] = params.metadata

        key_pair = KeyPairMaterial(
            algorithm=provider.algorithm,
            public_jwk=pub_rec["public_jwk"],
            private_jwk=priv_rec["private_jwk"],
        )
        data_bytes = cbor_encode(data)
        payload = cbor_encode([1, "tcert", data_bytes])
        bytes_out = build_signed_object("tcert", data, key_pair, provider)
        parsed = parse_signed_object(bytes_out)
        assert_valid_object_data("tcert", parsed.data)

        if not verify_parsed_signed_object(parsed, provider, pub_rec["public_jwk"]):
            raise QrsCryptoError("TCert self-signature verification failed")

        tcert_id = tcert_id_of(key_id, certificate_number)
        await self._deps.certificate_store.save(tcert_id, bytes_out)
        return CreateTcertResult(
            key_id=key_id,
            tcert_id=tcert_id,
            certificate_number=certificate_number,
            bytes=bytes_out,
            parsed=parsed,
        )

    async def get_tcert(self, tcert_id: str) -> tuple[bytes, Any]:
        data = await self._deps.certificate_store.get(tcert_id)
        if not data:
            raise QrsNotFoundError(f"TCert not found: {tcert_id}")
        return data, parse_signed_object(data)

    async def public_key_of(self, tcert_id: str) -> tuple[str, dict[str, Any]]:
        _, parsed = await self.get_tcert(tcert_id)
        return parsed.data["algorithm"], parsed.data["publicKey"]

File: src/qrs-core-js/src/services/certificateService.ts
/**
 * CertificateService: key generation and TCert creation.
 *
 * A TCert combines an issuer identity and the schema of one document type. It is
 * always self-signed, and its `key_id` is derived from its public key. A single key
 * pair may own several TCerts (one per document type), distinguished by their
 * certificate number.
 */
import type { PublicJwk } from '../crypto/jwk.js';
import { QrsCryptoError, QrsNotFoundError, QrsValidationError } from '../errors.js';
import { toHex } from '../id.js';
import { keyIdToBytes, tcertIdOf } from '../signedObject/signedObject.js';
import {
  buildSignedObject,
  parseSignedObject,
  verifyParsedSignedObject,
  type ParsedSignedObject,
} from '../signedObject/signedObject.js';
import { assertValidObjectData, isFieldType } from '../signedObject/schemas.js';
import { isHashAlgorithm, type HashAlgorithm } from '../id.js';
import type { AlgorithmId, KeyId, TcertId } from '../types.js';
import type { FieldSchema } from '../fields/types.js';
import type { ServiceDeps } from './deps.js';

export interface CreateTcertParams {
  algorithm: AlgorithmId;
  /** Human-readable name of this TCert. A TCert represents a document/entity;
   * the *issuer* display name comes from the CA that attests it (via trust). */
  name: string;
  /** The document schema: field definitions. */
  fields: FieldSchema[];
  /** Reuse an existing key; when omitted a new key pair is generated. */
  keyId?: KeyId;
  validAfter?: number;
  validBefore?: number;
  /**
   * Default validity duration (seconds) applied to every SDoc this TCert issues:
   * a document older than this at verification time is invalid. This lets a TCert
   * that is itself valid for a long time restrict how long its documents stay
   * valid (e.g. `sdocMaxAgeSeconds: 7 * 86_400` for one week).
   */
  sdocMaxAgeSeconds?: number;
  /** Hash algorithm for attachment content hashing under this TCert (default SHA-256). */
  hashAlgorithm?: HashAlgorithm;
  onlineEndpoint?: string;
  metadata?: Record<string, unknown>;
}

export interface CreateTcertResult {
  keyId: KeyId;
  tcertId: TcertId;
  certificateNumber: number;
  bytes: Uint8Array;
  parsed: ParsedSignedObject;
}

export class CertificateService {
  constructor(private readonly deps: ServiceDeps) {}

  /** Generate a new key pair and store it (private + public). Returns the key_id. */
  async generateKeyPair(algorithm: AlgorithmId): Promise<KeyId> {
    const provider = this.deps.cryptoRegistry.get(algorithm);
    const pair = await provider.generateKeyPair();
    const keyId = provider.keyId(pair.publicJwk);
    await this.deps.privateKeyStore.save(keyId, algorithm, pair.privateJwk);
    await this.deps.publicKeyStore.save(keyId, algorithm, pair.publicJwk);
    return keyId;
  }

  /** Create a self-signed TCert for a document type and store it. */
  async createTcert(params: CreateTcertParams): Promise<CreateTcertResult> {
    for (const field of params.fields) {
      if (!isFieldType(field.type)) throw new QrsValidationError(`Unsupported field type: ${field.type}`);
      if (!field.name || !field.label) throw new QrsValidationError('Each field must have a name and a label');
      if (!this.deps.fieldRegistry.has(field.type)) {
        throw new QrsValidationError(`No field engine registered for type: ${field.type}`);
      }
    }

    let keyId = params.keyId;
    if (!keyId) keyId = await this.generateKeyPair(params.algorithm);

    const pubRec = await this.deps.publicKeyStore.load(keyId);
    if (!pubRec) throw new QrsNotFoundError(`Public key not found: ${keyId}`);
    const privRec = await this.deps.privateKeyStore.load(keyId);
    if (!privRec) throw new QrsNotFoundError(`Private key not available for signing: ${keyId}`);

    const provider = this.deps.cryptoRegistry.get(pubRec.algorithm);

    // Pick the lowest unused certificate number (1..255) for this key.
    const existing = await this.deps.certificateStore.findByKeyId(keyId);
    const used = new Set(existing.map((e) => Number(e.tcertId.split(':')[1])));
    let certificateNumber = 0;
    for (let n = 1; n <= 255; n++) {
      if (!used.has(n)) {
        certificateNumber = n;
        break;
      }
    }
    if (certificateNumber === 0) throw new QrsValidationError('No certificate numbers left for this key');

    const data: Record<string, unknown> = {
      keyId: keyIdToBytes(keyId),
      certificateNumber,
      algorithm: provider.algorithm,
      publicKey: pubRec.publicJwk,
      identity: { name: params.name },
      schema: params.fields,
    };
    const validity: { validAfter?: number; validBefore?: number; sdocMaxAgeSeconds?: number } = {};
    if (params.validAfter !== undefined) validity.validAfter = params.validAfter;
    if (params.validBefore !== undefined) validity.validBefore = params.validBefore;
    if (params.sdocMaxAgeSeconds !== undefined) validity.sdocMaxAgeSeconds = params.sdocMaxAgeSeconds;
    if (Object.keys(validity).length > 0) data.validity = validity;
    if (params.hashAlgorithm !== undefined) {
      if (!isHashAlgorithm(params.hashAlgorithm)) {
        throw new QrsValidationError(`Unsupported hash algorithm: ${String(params.hashAlgorithm)}`);
      }
      data.hashAlgorithm = params.hashAlgorithm;
    }
    if (params.onlineEndpoint) data.onlineEndpoint = params.onlineEndpoint;
    if (params.metadata && Object.keys(params.metadata).length > 0) data.metadata = params.metadata;

    const keyPair = { algorithm: provider.algorithm, publicJwk: pubRec.publicJwk, privateJwk: privRec.privateJwk };
    const bytes = await buildSignedObject('tcert', data, keyPair, provider, new Uint8Array(0), certificateNumber);
    const parsed = parseSignedObject(bytes);
    assertValidObjectData('tcert', parsed.data);

    // Self-signature must hold, and the key_id must match the embedded public key.
    if (
      !(await verifyParsedSignedObject(parsed, provider, pubRec.publicJwk)) ||
      toHex(parsed.data.keyId as Uint8Array) !== keyId
    ) {
      throw new QrsCryptoError('TCert self-signature verification failed');
    }

    const tcertId = tcertIdOf(keyId, certificateNumber);
    await this.deps.certificateStore.save(tcertId, bytes);
    return { keyId, tcertId, certificateNumber, bytes, parsed };
  }

  /** Convenience: load and parse a stored TCert. */
  async getTcert(tcertId: TcertId): Promise<{ bytes: Uint8Array; parsed: ParsedSignedObject }> {
    const bytes = await this.deps.certificateStore.get(tcertId);
    if (!bytes) throw new QrsNotFoundError(`TCert not found: ${tcertId}`);
    return { bytes, parsed: parseSignedObject(bytes) };
  }

  /** Convenience: derive the public JWK embedded in a TCert. */
  async publicKeyOf(tcertId: TcertId): Promise<{ algorithm: AlgorithmId; publicJwk: PublicJwk }> {
    const { parsed } = await this.getTcert(tcertId);
    const algorithm = parsed.data.algorithm as AlgorithmId;
    const publicJwk = parsed.data.publicKey as unknown as PublicJwk;
    return { algorithm, publicJwk };
  }
}
